src/ion_writer.rs

```diff
@@ -97,29 +97,38 @@ pub trait IonWriter {
     /// This method can only be called when the writer is at the top level.
     fn flush(&mut self) -> IonResult<()>;
 
-    /// Returns a reference to the writer's output.
-    ///
-    /// This method can be used to inspect the Ion data that the writer has produced without having
-    /// to first drop the writer.
-    /// ```
-    /// use ion_rs::element::Element;
-    /// use ion_rs::{IonResult, IonWriter, TextWriter, TextWriterBuilder};
-    /// # fn roundtrip() -> IonResult<()> {
-    /// // Set up our output buffer
-    /// let mut buffer: Vec<u8> = Vec::new();
-    /// // Construct a writer that will serialize values to the buffer
-    /// let mut writer = TextWriterBuilder::default().build(&mut buffer)?;
-    /// // Serialize some data
-    /// writer.write_string("hello")?;
-    /// writer.flush()?;
-    /// // Read the data back from the output buffer
-    /// let output_element = Element::read_one(writer.output())?;
-    /// // Confirm that it matches the input data
-    /// assert_eq!(Element::from("hello"), output_element);
-    /// # Ok(())
-    /// # }
-    /// # roundtrip().unwrap();
-    /// ```
+    // This portion of the doc comment is always visible
+    #[doc = r##"
+Returns a reference to the writer's output.
+
+This method can be used to inspect the Ion data that the writer has produced without
+having to first drop the writer.
+    "##]
+    // This part is only visible if the "experimental-writer" feature is enabled.
```

Comment:
🗺️ This is a pair of doc comments using [an unfamiliar syntax](https://users.rust-lang.org/t/doctests-that-require-a-non-default-feature-is-it-possible/29529/7); it allows parts of the doc comment to be hidden/shown depending on the features that are enabled.

Reply:
It's cool, but why is it necessary here?

Reply:
A confluence of a few factors:
* The `IonWriter` code gets compiled regardless of whether the `"experimental-writer"` feature is enabled. It's used to implement always-available public APIs like `Element::read_one()`.
* Doc examples can only use public APIs, so if we have a doc example that refers to an API that isn't available, the doc test will fail.
* We always build/run the doc tests on our private APIs as part of our CI. This guarantees that references to types and traits mentioned in the comments remains correct/intact.

To rephrase: if you build without `"experimental-writer"`, `IonWriter` and its doc tests still get built/run; a doc test referring to an unavailable API like `TextWriterBuilder` will cause a CI failure.